EasyCommands/CommandParsers/ProcessingEngine.cs

```diff
@@ -154,6 +154,11 @@ partial class Program {
             OneValueRule(Type<UniOperationCommandParameter>, requiredRight<VariableCommandParameter>(),
                 (p, df) => new VariableCommandParameter(new UniOperandVariable(p.value, df.value))),
 
+            //DynamicVectorProcessor
+            FourValueRule(Type<TernaryConditionSeparatorParameter>, requiredLeft<VariableCommandParameter>(), requiredRight<VariableCommandParameter>(), requiredRight<TernaryConditionSeparatorParameter>(), requiredRight<VariableCommandParameter>(),
+                (sep1, x, y, sep2, z) => AllSatisfied(x, y, z) && !(x.GetValue().value is VectorVariable || y.GetValue().value is VectorVariable || z.GetValue().value is VectorVariable),
```

Comment:
You can't use x.Getvalue().value here, as during parsing you don't actually know the value of the variable.

I think you simply look for the pattern var:var:var and treat that as a dynamic VectorVariable.

Only at runtime can you actually determine.  In that code I wouldn't do too much.  A simple CastNumber(v) should be sufficient.  Perhaps we can wrap it with a try catch to throw a clear "invalid variable used in vector creation" message or similar.

Reply:
This only resolves down to `Variable` not to its primitive value.
That pattern is repeatedly used to check the actual type of the `Variable`.

Reply:
The problem is that is not parsed that way without the constrained.

We have the rules
```
vector:  var : var : var -> var
ternary: var ? var : var -> var
```

Let us consider there was no constraint on the vector rule.
If we try to parse `var? var:var:var : var:var:var` with these two rules we get the follwoing sequence (token | rule, cursor) where indices are added to the variables to distinguish them.
```
v1 ? v2 : v3 : v4 : v5 : v6 : v7 | vector
                       ^
v1 ? v2 : v3 : v4 : v8 | vector
             ^
v1 ? v2 : v9 | ternary
   ^
v10
```
`v9` is an invalid vector since its z component is `v8`, another  vector.

With the constraint that the vector rule cannot consume vector variables we get the desired sequence
```
v1 ? v2 : v3 : v4 : v5 : v6 : v7 | vector'
                       ^
v1 ? v2 : v3 : v4 : v8 | vector' 
         ^
v1 ? v9 : v8 | ternary
   ^
v10
```

Reply:
GPS would be even worse since i dont see a way to constrain the rule so it does not get applied when we actually mean vectors or we are in a similar situation as above or even worse are in situation where the vector rule tries to consume a gps variable.

Reply:
I have a crazy idea.

I propose that we introduce "::" as another TernaryOperationSeparator indicator, added to first pass tokens.

Then, during string parsing, we look for " : " (specifically looking for spaces) and convert to " :: ".

Then, the rules only look for AmbiguousColonSeparator for processing vectors or gps coordinates.

Vectors or gps coordinates should never be defined with spaces (we can even mention this). a : b : c  would not be processed as a vector, ever.  You must specify a:b:c (not unreasonable).

After the GPS/Vector parsing rules, we add a simple rule that converts AmbiguousColonSeparator to TernaryOperationSeperator

This approach, I believe would yield correct results, provided the user spaces correctly.

set a to b ? c:d:e : a:0:0:0:0 would be processed correctly with this approach.  The only ambiguous case would be if the user tries set a to b?c:d:e:a:0:0:0:0, which I don't think will happen because it's pretty obvious, even to the user, that it's not clear what they were trying to do.

This approach gets rid of the need for VectorVariable checking in the GPS and Vector Processing rules as well.  You can always do GPS first, and Vector second.  It also means that a user can further clarify the ternary operation intent by explicitly specifying "::" instead of ":".  We could also consider ";".

Thoughts?

Reply:
I just sent out a PR for adding that behavior, which can be pushed as is if you want to rebase.  This should get you the separation desired so you can do some simple vector & gps parsing rules.

https://github.com/MerlinofMines/EasyCommands/pull/173/files

Reply:
I do not think this a good idea. It enables the user again to indiscriminately but other vectors variables into a vectors.
In the current state they have to atleast hide the vector behind an operation.

I put a few thoughts into the GPS issue. A GPS marker is a __string__ of the form `GPS:[name]:[X]:[Y]:[Z]:[color]`. 
_Note the leading `GPS`_. This can be unambigously matched with a 10 value rule or 3 rules to partially build it up.

__But__ there is actually no point to pretend that it is anything but a string since the block interface wants a string anyway. Casting such a string to vector already works (for whatever reason) and there is no use case for knowing the name or the color of a GPS marker, so a cast in the other direction is pointless. Then only some handling for properties that want a GPS (string) is needed to allow vector input and convert it into the desired format.

Reply:
I think we may actually need both.

As is the VectorVariable alone does not disambiguate ```a ? b : c:d:e``` from ```a ? b:c:d : e```.  For that, my pull request is needed.

The VectorVariable check helps with ```a?b:c:d:e:f:g```, interpreted as a ? b:c:d e:f:g.  I'm not even sure how reasonable it is to expect b:c:d:e:f:g to be interpreted correctly, TBD.  Most likely, a user will add spaces to clarify, and I think the parsing should use that information to help parse meaning.

I'm happy to keep the VectorVariable check as a way to protect users against themself, but I'm more concerned about handling cases where to the user it's clear what they're trying to do but it's ambiguous to the script parsing because we lost context on association via spacing.

I also don't think VectorVariable needs to be applied to

**GPS Parsing**

We can certainly solve GPS parsing separately, let's separate it from this pull request and focus on dynamic vectors instead.

Reply:
lets continue the discussion threads in the main thread below